src/jaxoplanet/_src/experimental/starry/rotation.py

```diff
@@ -0,0 +1,105 @@
+from typing import Callable
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+from jax.scipy.linalg import block_diag
+from scipy.special import factorial
+
+from jaxoplanet._src.types import Array
+
+
+@jax.jit
+def axis_to_euler(u1: float, u2: float, u3: float, theta: float):
+    """Axis-angle rotation matrix"""
+    tol = 1e-15
+    cos_theta = jnp.cos(theta)
+    sin_theta = jnp.sin(theta)
+    P01 = u1 * u2 * (1 - cos_theta) - u3 * sin_theta
+    P02 = u1 * u3 * (1 - cos_theta) + u2 * sin_theta
+    P11 = cos_theta + u2 * u2 * (1 - cos_theta)
+    P12 = u2 * u3 * (1 - cos_theta) - u1 * sin_theta
+    P20 = u3 * u1 * (1 - cos_theta) - u2 * sin_theta
+    P21 = u3 * u2 * (1 - cos_theta) + u1 * sin_theta
+    P22 = cos_theta + u3 * u3 * (1 - cos_theta)
+    norm1 = jnp.sqrt(P20 * P20 + P21 * P21)
+    norm2 = jnp.sqrt(P02 * P02 + P12 * P12)
+
+    alpha0 = jnp.arctan2(0.0, 1.0)
+    beta0 = jnp.arctan2(0.0, -1.0)
+    gamma0 = jnp.arctan2(P01, P11)
+
+    alpha1 = jnp.arctan2(0.0, 1.0)
+    beta1 = jnp.arctan2(0.0, 1.0)
+    gamma1 = jnp.arctan2(-P01, P11)
+
+    alpha2 = jnp.arctan2(P12 / norm2, P02 / norm2)
+    beta2 = jnp.arctan2(jnp.sqrt(1 - P22**2), P22)
+    gamma2 = jnp.arctan2(P21 / norm1, -P20 / norm1)
+
+    case1 = jnp.logical_and((P22 < -1 + tol), (P22 > -1 - tol))
+    case2 = jnp.logical_and((P22 < 1 + tol), (P22 > 1 - tol))
+
+    alpha = jnp.where(case1, alpha0, jnp.where(case2, alpha1, alpha2))
+    beta = jnp.where(case1, beta0, jnp.where(case2, beta1, beta2))
+    gamma = jnp.where(case1, gamma0, jnp.where(case2, gamma1, gamma2))
```

Comment:
Around these lines I think we'll want to add a test that the gradients don't return `nan`. If the gradient of one branch of the `where` is `nan`, even if it's not used, the `where`'s gradient becomes `nan`.

Reply:
I see. Should that be tested in a global test (like `~jnp.isnan(grad(axis_to_euler))`? Or in-line tests in the function to ensure a finite return? (first time with jax autodiff features)

Reply:
I was thinking a test case - inline tests wouldn't really do the trick. Choose some inputs that you know will result in NaNs for some branches and test that the gradients are finite!